File: 1 Algorithmic Toolbox/Week 5 Dynamic Programming 1/week5_dynamic_programming1/1_money_change_again/change_dp.py
# Uses python3
import sys
from math import inf

def get_change(m):
    #write your code here
    coins = [1, 3, 4]
    MinNumCoins = [0] * (m + 1)
    
    for i in range(1, m+1):
        MinNumCoins[i] = inf
        for coin in coins:
            if i >= coin:
                NumCoins = MinNumCoins[i-coin] + 1
                if NumCoins < MinNumCoins[i]:
                    MinNumCoins[i] = NumCoins
     
    return MinNumCoins[m]

if __name__ == '__main__':
    m = int(sys.stdin.read())
    print(get_change(m))

File: 1 Algorithmic Toolbox/Week 5 Dynamic Programming 1/week5_dynamic_programming1/1_money_change_again/change_dp.cpp
#include <iostream>
#include <limits>
#include <vector>

using namespace std;
int INF = numeric_limits<int>::max();

int get_change(int m) {
  //write your code here
  vector<int> coins = {1, 3, 4};
  
  vector<int> MinNumCoins(m+1);
  
  for (int i = 1; i < m+1; i++){
      MinNumCoins[i] = INF;
      for (int j = 0; j < coins.size(); j++){
          if (i >= coins[j]){
              int NumCoins = MinNumCoins[i-coins[j]] + 1;
              if (NumCoins < MinNumCoins[i]){
                  MinNumCoins[i] = NumCoins;
              }
          }
      }
  
  }
  
  return MinNumCoins[m];
}

int main() {
  int m;
  std::cin >> m;
  std::cout << get_change(m) << '\n';
}
